HW2/P3/mandelbrot.pyx

```diff
@@ -26,23 +26,66 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
        #     v3 = AVX.mul(v1, v2)
        #
        # You may find the numpy.real() and numpy.imag() fuctions helpful.
+       AVX.float8 c_r, c_i, z_r, z_r_tmp, z_i, results, ones, fours, filter
 
     assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
     assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
     assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
 
     with nogil:
-        for i in range(in_coords.shape[0]):
-            for j in range(in_coords.shape[1]):
-                c = in_coords[i, j]
-                z = 0
+        for i in prange(in_coords.shape[0], schedule='static', chunksize=1, num_threads=4):
+            # for j in range(in_coords.shape[1]):
+            #     c = in_coords[i, j]
+            #     z = 0
+
+            for j in range(0, in_coords.shape[1], 8):
+                c_r = AVX.make_float8(in_coords[i, j+7].real,
+                                      in_coords[i, j+6].real,
+                                      in_coords[i, j+5].real,
+                                      in_coords[i, j+4].real,
+                                      in_coords[i, j+3].real,
+                                      in_coords[i, j+2].real,
+                                      in_coords[i, j+1].real,
+                                      in_coords[i, j].real)
+                c_i = AVX.make_float8(in_coords[i, j+7].imag,
+                                      in_coords[i, j+6].imag,
+                                      in_coords[i, j+5].imag,
+                                      in_coords[i, j+4].imag,
+                                      in_coords[i, j+3].imag,
+                                      in_coords[i, j+2].imag,
+                                      in_coords[i, j+1].imag,
+                                      in_coords[i, j].imag)
+                fours = AVX.float_to_float8(4.0)
+                ones = AVX.float_to_float8(1.0)
+                z_r = AVX.float_to_float8(0.0)
+                z_i = AVX.float_to_float8(0.0)
+                results = AVX.float_to_float8(0.0)
+
                 for iter in range(max_iterations):
-                    if magnitude_squared(z) > 4:
+
+
+                    # if magnitude_squared(z) > 4:
+
+
+                    filter = AVX.greater_than(fours, AVX.fmadd(z_r, z_r, AVX.mul(z_i, z_i)))
+                    if AVX.signs(filter) == 0:
                         break
-                    z = z * z + c
-                out_counts[i, j] = iter
 
 
+                    # z = z * z + c
+
+
+                    results = AVX.add(results, AVX.bitwise_and(filter, ones))
+                    z_r_tmp = AVX.fmsub(z_r, z_r, AVX.mul(z_i, z_i))
+                    z_i = AVX.fmadd(z_r, z_i, AVX.mul(z_i, z_r))
+                    z_i = AVX.add(z_i, c_i)
+                    z_r = AVX.add(z_r_tmp, c_r)
```

Comment:
Looks good. I like how you compressed most of the operations together. From my understanding, although more difficult to read, this may improve the performance when compiling.